app/src/main/java/protect/card_locker/importexport/CatimaImporter.java

```diff
@@ -368,22 +487,57 @@ private void importLoyaltyCard(Context context, SQLiteDatabase database, CSVReco
+    private void importGroupV2(SQLiteDatabase database, CSVRecord record) throws FormatException {
+        String name = CSVHelpers.extractString(DBHelper.LoyaltyCardDbGroups.ID, record, null);
+        DBHelper.insertGroup(database, name);
+    }
 
-        DBHelper.insertGroup(database, id);
+    /**
+     * Import a single group from the V3 scheme (database v16) into the database using the given
+     * session.
+     */
+    private Hashtable<Integer, String> importGroupV3(SQLiteDatabase database, List<CSVRecord> records) throws FormatException {
+        Hashtable<Integer,String> groupsTable = new Hashtable<>();
+        TreeMap<Integer, Group> sortedGroups = new TreeMap<Integer, Group>();
+        for (CSVRecord record : records) {
+            Integer id = CSVHelpers.extractInt(DBHelper.LoyaltyCardDbGroups.ID, record, false);
+            String name = CSVHelpers.extractString(DBHelper.LoyaltyCardDbGroups.NAME, record, null);
+            Integer order = CSVHelpers.extractInt(DBHelper.LoyaltyCardDbGroups.ORDER, record, false);
+            DBHelper.insertGroup(database, name);
+            sortedGroups.put(order, DBHelper.getGroupByName(database, name));
+            groupsTable.put(id, name);
+        }
+        List<Group> sortedGroupList = new ArrayList<>(sortedGroups.values());
+        DBHelper.reorderGroups(database, sortedGroupList);
+        return groupsTable;
     }
 
     /**
      * Import a single card to group mapping into the database using the given
      * session.
      */
-    private void importCardGroupMapping(SQLiteDatabase database, CSVRecord record) throws FormatException {
+    private void importCardGroupMappingV2(SQLiteDatabase database, CSVRecord record) throws FormatException {
         Integer cardId = CSVHelpers.extractInt(DBHelper.LoyaltyCardDbIdsGroups.cardID, record, false);
-        String groupId = CSVHelpers.extractString(DBHelper.LoyaltyCardDbIdsGroups.groupID, record, null);
+        List<Group> cardGroups = DBHelper.getLoyaltyCardGroups(database, cardId);
+
+        String groupName = CSVHelpers.extractString(DBHelper.LoyaltyCardDbIdsGroups.groupID, record, null);
+        cardGroups.add(DBHelper.getGroupByName(database, groupName));
+
+        DBHelper.setLoyaltyCardGroups(database, cardId, cardGroups);
+    }
 
+    /**
+     * Import a single card to group mapping from V3 scheme (database v16) into the database using the given
+     * session.
+     */
+    private void importCardGroupMappingV3(SQLiteDatabase database, CSVRecord record, Hashtable<Integer,String> groupsTable) throws FormatException {
+        Integer cardId = CSVHelpers.extractInt(DBHelper.LoyaltyCardDbIdsGroups.cardID, record, false);
         List<Group> cardGroups = DBHelper.getLoyaltyCardGroups(database, cardId);
+
+        Integer groupId = CSVHelpers.extractInt(DBHelper.LoyaltyCardDbIdsGroups.groupID, record, false);
         cardGroups.add(DBHelper.getGroup(database, groupId));
```

Comment:
Not 100% sure this is correct, but it makes the tests pass.

```suggestion
        cardGroups.add(DBHelper.getGroupByName(database, groupsTable.get(groupId)));
```

Reply:
Also not sure how importing works. Does it keep existing cards and groups? If so does this handle conflicting IDs between existing data and the exported data?

Reply:
Might be better to have an `DBHelper.insertGroup(database, id, name)` to be used in `importGroupV3()` to insert a group with a specific ID.